### PR TITLE
graphics/deko3d: fix and simplify rbtree code

### PR DESCRIPTION
**Pull request by scturtle**
Solve #100.

### CODE REVIEW

File: graphics/deko3d/deko_examples/source/SampleFramework/CIntrusiveTree.h [hunk 1]
```diff
@@ -129,13 +129,7 @@ class CIntrusiveTreeBase
...
 constexpr ClassT* parent_obj(MemberT* member, MemberT ClassT::* ptr)
 {
-    union whatever
-    {
-        MemberT ClassT::* ptr;
-        intptr_t offset;
-    };
-    // This is technically UB, but basically every compiler worth using admits it as an extension
-    return (ClassT*)((intptr_t)member - whatever{ptr}.offset);
+    return (ClassT *)((char *)member - (char *)&((ClassT *)0->*ptr));
```
Comment (fincs) [inline graphics/deko3d/deko_examples/source/SampleFramework/CIntrusiveTree.h:132]: This is still UB fwiw. I'd say the original version of this function is easier to read.

File: graphics/deko3d/deko_examples/source/SampleFramework/CIntrusiveTree.h [hunk 2]
```diff
@@ -194,31 +188,11 @@ class CIntrusiveTree final : protected CIntrusiveTreeBase
...
-                        node = walk(parent, N::Right);
-                    else
-                        node = parent;
-                }
-                break;
+        if (mode != Exact) {
+            if (mode == UpperBound && node)
+                node = walk(node, N::Right);
+            else if (!node && parent)
+                node = &parent->left() == &point ? parent : walk(parent, N::Right);
```
Comment (fincs) [inline graphics/deko3d/deko_examples/source/SampleFramework/CIntrusiveTree.h:195]: Is this ternary actually correct for UpperBound mode?

File: graphics/deko3d/deko_examples/source/SampleFramework/CIntrusiveTree.h (hunk 1 above)
Comment (scturtle) [reply graphics/deko3d/deko_examples/source/SampleFramework/CIntrusiveTree.h]: Reverted.

File: graphics/deko3d/deko_examples/source/SampleFramework/CIntrusiveTree.h (hunk 2 above)
Comment (scturtle) [reply graphics/deko3d/deko_examples/source/SampleFramework/CIntrusiveTree.h:201]: Please let me know if there is any problem.

Comment (fincs) [reply graphics/deko3d/deko_examples/source/SampleFramework/CIntrusiveTree.h:201]: I was mostly asking how did you arrive to that simplification, as I'm currently struggling to wrap my head around this code :p

Comment (fincs) [reply graphics/deko3d/deko_examples/source/SampleFramework/CIntrusiveTree.h:201]: After some analysis, I believe this code to be correct.

**Pull request merged**